vm/src/builtins.rs

```diff
@@ -559,32 +559,60 @@ pub struct PrintOptions {
     end: Option<PyStringRef>,
     #[pyarg(keyword_only, default = "false")]
     flush: bool,
+    #[pyarg(keyword_only, default = "None")]
+    file: Option<PyObjectRef>,
 }
 
 pub fn builtin_print(objects: Args, options: PrintOptions, vm: &VirtualMachine) -> PyResult<()> {
-    let stdout = io::stdout();
-    let mut stdout_lock = stdout.lock();
-    let mut first = true;
-    for object in objects {
-        if first {
-            first = false;
-        } else if let Some(ref sep) = options.sep {
-            write!(stdout_lock, "{}", sep.value).unwrap();
-        } else {
-            write!(stdout_lock, " ").unwrap();
+    if let Some(file) = options.file {
+        let sep = match options.sep {
+            Some(sep) => sep.into_pyobject(vm).unwrap(),
+            _ => PyString::from(" ").into_pyobject(vm).unwrap(),
+        };
+
+        let mut first = true;
+        for object in objects {
```

Comment:
This for loop is now duplicated, including the `first` variable. I would change this into something like this:

Enter pseudo rust code:
```rust

// Create some sort of closure function:
let write_func = if let Some(file) = options.file {
  get_
} else {
  func_with_write_macro
}

// From this point, we have the availability over a function which does the proper printing, and can use this function at will.

// Now loop over the objects and print them using the closure function:
let mut first = true;
for object in objects {
  
  if first {
    first = false;
  } else {
    write_func(sep.clone())
  }
  write_func(object)
}
```

Reply:
Additional information: https://doc.rust-lang.org/book/ch13-01-closures.html

Reply:
better now?

Reply:
Nice job!